### ISSUE TITLE
s param in glmnet

### ISSUE THREAD

**Issue by Coorsaa** [mlr-org/mlr3learners#10]
glmnet predicts a matrix for different values for `s`. Hence one need to specify a certain value for it to get a single prediction vector. in mlr and right now, we set s to 0.01 and dismiss all 100 calculated values. How should we deal with this? switch to cvglmnet?

**Comment by mllg** [comment]
See https://github.com/mlr-org/mlr3/issues/273.

**Comment by pat-s** [comment]
> in mlr and right now, we set s to 0.01 and dismiss all 100 calculated values.

Learner `glmnet` is dangerous if one does not know how it works. It requires manual tuning of `s` and simply using the default from our side makes no sense at all.

I'd argue that it would be more safe to require user input for `s` or error otherwise rather than using a nonsense default of 0.01 (like in mlr2) just to get it working. Also we should mention that `lambda` has no relevance for `glmnet::glmnet()` if a custom `s` is passed.

In mlr3 we currently do not use `glmnet::glmnet()` but only `glmnet::cvglmnet()` which always uses `lambda.1se`, optimized during an internal 10-fold CV, for `s`.

We should be more verbose with all of this in the learner notes:

- What happens if the user sets something for `s` (custom tuning required)
- What happens if multiple values for `s` are supplied.

TBH, I do not see a use-case for supplying multiple `s` values in any case - I'd limit it internally to 1L and require model refitting for every `s` supplied.

**Comment by berndbischl** [comment]
> Learner `glmnet` is dangerous if one does not know how it works. It requires manual tuning of `s` and simply using the default from our side makes no sense at all.

but thats exactly the same "problem" for (m)any regularized learners? svms?

**Comment by berndbischl** [comment]
theres lots of instances of these types of problems. we have ressource-problems if we handle them again and again indivifually

**Comment by pat-s** [comment]
> but thats exactly the same "problem" for (m)any regularized learners? svms?

That's true but there is a small difference between let's say `C` of a SVM and `s` in `glmnet`: All packages of the former do not perform any automated optimization.

For {glmnet} fitting "just works" due to the internal heuristics of the package. So users are not used how to handle {glmnet} in many cases (I see that in practice here a lot). Then there is a lot of confusion between the difference of `s` and `lambda` (yes, not our problem).

However if we set a custom default for `s` (default in `predict.glmnet()` is NULL) then we should do that with some meaning? By this I mean a sensible default like `num.tress` for RF. 
However, this does not exist for {glmnet} and hence I'd argue we should set no default (or just use the actual default for which the user would notice what would be required).

I used to tune `s` in mlr2 via MBO and I am missing the option to do with mlr3 because we are currently fixed to `cv.glmnet`.

I'd expose the choice of whether to use `cv.glmnet` or `glmnet` as a non-tuneable hyperpar. `s` would be a dep param on setting `glmnet`. What do you think?

**Comment by berndbischl** [comment]
fine, sounds good

**Comment by mllg** [comment]
I disagree. 

Tuning `s` is much more efficient if done internally (at least it should, from the theory).

Additionally, there are many hyperparameters which **must** be tuned to achieve competitive performance. Another example is `nrounds` of `xgboost`. This is a problem we cannot solve. I don't see a difference if we set some defaults for technical reasons or the package developer sets some defaults for technical reasons.

**Comment by pat-s** [comment]
> Tuning s is much more efficient if done internally (at least it should, from the theory).

I am not debating which approach is better. I just state that there are two and how to handle `s` in the best way.

In addition we should not decide for the user that one cannot use `glmnet`/tune manually.

Especially in a benchmark comparison you want to rely on the same tuning method for consistency and not rely on the internal tuning by {glmnet}.

FWIW I do exactly this in my current paper and see no performance loss tuning (with MBO in this case) compared to the internal one of {glmnet}.

Note that I also do not want to change any defaults here: I simply want to add the option to use `glmnet()` instead of `cv.glmnet()` and document the role of `s` in all of that. In a single learner, with the option to choose one as a "backend".

The solution we had in mlr2 with two distinct learners for the same algorithm implementation with a nonsense default was not optimal imo.

>  I don't see a difference if we set some defaults for technical reasons or the package developer sets some defaults for technical reasons.

Well, in this case we create the "next version" of this learner with different defaults. Instead our goal is/should be to just provide the learners in a consistent interface?

> Additionally, there are many hyperparameters which must be tuned to achieve competitive performance. 

Again to make that clear: This is not my point here. I just want to provide the option to tune `s` at all and document its behavior. And **if** we change a default, than we should not change it to a "works for us because of technical reasons but does not make sense at all".
